viahtml/blocklist.py

```diff
@@ -46,15 +46,45 @@ def __init__(self, filename):
         self._filename = filename
         self._last_modified = None
         self.domains = {}
+        self.patterns = {}
 
         self._refresh()
 
     def is_blocked(self, url):
-        """Get the blocked reason for a URL, or None if it's not blocked."""
+        """Get the blocked reason for a URL, or False if it's not blocked."""
         self._refresh()
 
         domain = self._domain(url)
-        return self.domains.get(domain, False)
+        blocked = self.domains.get(domain)
+        if blocked:
+            return blocked
+
+        for pattern, reason in self.patterns.items():
+            if pattern.match(domain):
+                return reason
+
+        return False
+
+    def clear(self):
+        """Remove all domains from the blocklist."""
+        self.domains, self.patterns = {}, {}
+
+    def add_domain(self, domain, reason):
+        """Add a domain (or domain pattern) to the blocklist."""
+
+        reason = Blocklist.Reason.parse(reason)
+        if reason in self.PERMITTED:
+            # This is listed as blocked, but this service can actually
+            # serve this type without incident
+            return
+
+        if "*" in domain:
+            # Convert a string with '*' wildcards into a regex
+            pattern = "^" + re.escape(domain).replace("\\*", ".*") + "$"
+            pattern = re.compile(pattern, re.IGNORECASE)
+            self.patterns[pattern] = reason
```

Comment:
This is the relevant change for the most part

Reply:
I think you could use fnmatch, which supports `*`, `?` and `[]` (character ranges) to avoid having to translate into regexes: https://docs.python.org/3/library/fnmatch.html#module-fnmatch

Reply:
I've used `fnmatch` here to produce the regex, which lets us have our cake and not code it. We could blindly apply `fnmatch` to all of them and simplify the code but:

 * For non wildcard things we can do a direct lookup, which is nearly constant time
 * I'm not sure if this is case insensitive, and we need it to be
 * I've already written it this way and I'm colossally lazy
 * This should be faster, as we compile the pattern once, rather than very time inside `fnmatch`